File: ts_webgl/src/geometry/Vector2.ts
export class Vector2 {
	constructor(
		readonly x: number,
		readonly y: number,
	) { }

	get negated(): Vector2 {
		return new Vector2(
			-this.x,
			-this.y,
		);
	}

	add(other: Vector2): Vector2 {
		return new Vector2(
			this.x + other.x,
			this.y + other.y,
		);
	}

	sub(other: Vector2): Vector2 {
		return new Vector2(
			this.x - other.x,
			this.y - other.y,
		);
	}

	mul(other: number): Vector2 {
		return new Vector2(
			this.x * other,
			this.y * other,
		);
	}

	div(other: number): Vector2 {
		return new Vector2(
			this.x / other,
			this.y / other,
		);
	}

	get magnitudeSquared(): number {
		return this.x * this.x + this.y * this.y;
	}

	get magnitude(): number {
		return Math.sqrt(this.magnitudeSquared);
	}

	get normalized(): Vector2 {
		return this.div(this.magnitude);
	}

	dot(other: Vector2): number {
		return this.x * other.x + this.y * other.y;
	}
}

File: rust-webgl/lib/src/glmath/vector2.rs
use std::{
    fmt::{Debug, Display},
    ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Neg, Sub, SubAssign},
};

use super::numbers::Float;

#[repr(C)]
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Vector2<T> {
    pub x: T,
    pub y: T,
}

impl<T> Vector2<T> {
    pub fn new(x: T, y: T) -> Self {
        Self { x, y }
    }
}

impl<T> Display for Vector2<T>
where
    T: Display,
{
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "({}, {})", self.x, self.y)
    }
}

impl<T> Vector2<T>
where
    T: Float + Copy,
{
    pub fn magnitude_squared(self) -> T {
        self.x * self.x + self.y * self.y
    }

    pub fn magnitude(self) -> T {
        self.magnitude_squared().sqrt()
    }

    pub fn normalized(self) -> Vector2<T> {
        self / self.magnitude()
    }

    pub fn dot_product(self, rhs: Self) -> T {
        self.x * rhs.x + self.y * rhs.y
    }
}

impl<T> Add for Vector2<T>
where
    T: Float + Copy,
{
    type Output = Self;

    fn add(self, rhs: Self) -> Self::Output {
        Self::Output {
            x: self.x + rhs.x,
            y: self.y + rhs.y,
        }
    }
}

impl<T> AddAssign for Vector2<T>
where
    T: Float + Copy,
{
    fn add_assign(&mut self, rhs: Self) {
        *self = *self + rhs
    }
}

impl<T> Sub for Vector2<T>
where
    T: Float + Copy,
{
    type Output = Self;

    fn sub(self, rhs: Self) -> Self::Output {
        Self::Output {
            x: self.x - rhs.x,
            y: self.y - rhs.y,
        }
    }
}

impl<T> SubAssign for Vector2<T>
where
    T: Float + Copy,
{
    fn sub_assign(&mut self, rhs: Self) {
        *self = *self + rhs
    }
}

impl<T> Mul<T> for Vector2<T>
where
    T: Float + Copy,
{
    type Output = Self;

    fn mul(self, rhs: T) -> Self::Output {
        Self::Output {
            x: self.x * rhs,
            y: self.y * rhs,
        }
    }
}

impl<T> MulAssign<T> for Vector2<T>
where
    T: Float + Copy,
{
    fn mul_assign(&mut self, rhs: T) {
        *self = *self * rhs
    }
}

impl<T> Div<T> for Vector2<T>
where
    T: Float + Copy,
{
    type Output = Self;

    fn div(self, rhs: T) -> Self::Output {
        Self::Output {
            x: self.x / rhs,
            y: self.y / rhs,
        }
    }
}

impl<T> DivAssign<T> for Vector2<T>
where
    T: Float + Copy,
{
    fn div_assign(&mut self, rhs: T) {
        *self = *self / rhs
    }
}

impl<T> Neg for Vector2<T>
where
    T: Float + Copy,
{
    type Output = Self;

    fn neg(self) -> Self::Output {
        Self::Output {
            x: -self.x,
            y: -self.y,
        }
    }
}
